### ISSUE TITLE
Using the Generator as a View Helper

### ISSUE THREAD

**Issue by jakejohns** [auraphp/Aura.Router#114]
I want to use the `Generator` in a view script.
It seems like I want to be able to do it like this, I guess?

``` php
// view.php
echo $this->a(
    $this->url()->generate('route', $params),
    'click here'
);
// or possibly even:
echo $this->a(
    $this->url('route', $params),
    'click here'
);
```

I did it like this the other day, and it seemed a little awkward.
I set up my DI to grab the Aura\Html helper locator, and set the `url` key to a
`lazyGetCall` on the router, and then wrapped that in a function...

``` php
<?php
class Config extends ContainerConfig
{
    public function modify(Container $di)
    {
        $helpers = $di->get('aura/html:helpers');
        $url = $di->lazyGetCall('radar/adr:router', 'getGenerator');
        $helpers->set(
            'url',
            function () use ($url) {
                return $url;
            }
        );
    }
}
```

...That seems pretty weird though.

If `Generator` had this:

``` php
//...
public function __invoke($name = null, $data = [])
{
    if (null == $name) {
        return $this;
    }

    return $this->generate($name, $data);
}
```

I wouldn't need to wrap it in a function to use with the helpers, and I could
just call `$this->url(...)`, instead of
`$this->url()->generate(...)`, but if needed, could also
`$this->url()->generateRaw(...)`

However, again... it doesnt seem quite right to me to add the `__invoke` method
like that in the `Generator`.

Maybe I should just be adding it into the view and not mess with the helpers?
eg:

``` php
<?php
class Config extends ContainerConfig
{
    public function modify(Container $di)
    {
        $view = $di->get('aura/view:view');
        $generator = $di->get('radar/adr:router')->getGenerator();
        $view->addData(['urlHelper' => $generator]);
    }
}

// in view:
echo $this->a(
    $this->urlHelper->generate('route', $params),
    'click here'
);
```

Are other people doing this?
Thoughts? Recommendations? Best Practices?

**Comment by dlundgren** [comment]
You could also decorate the generator and inject that into the view.

I've injected Aura.Router into a custom view helper to help keep url functionality together (generating routes, retrieving the base url, etc). It also allowed me to test that the view helper was doing what I expected. In some cases I need full urls, in others relative urls are fine.

**Comment by jakejohns** [comment]
I had been working on decorating the generator, and thought it a bit gratuitous, but perhaps my initial impulse was correct. @dlundgren, got a package, or any public code? I'd be curious to see your solution concretely.  I'm also curious if the "etc" in your list of URL functionality includes something like a mechanism for determining active state of a link. Right now, my responders are setting the PSR7 Request object to the view and I'm doing things like this for navigation:

``` php
foreach ($items as $name => $route) {
    $url = $this->urlHelper->generate($route);
    $attr = $this->request->getUri()->getPath() == $url ? ['class' => 'active'] : [];
    $menu->rawItem($this->anchor($url, $name), $attr);
}
```

**Comment by harikt** [comment]
I did it for v2 . You can see the sample code of view helper https://github.com/friendsofaura/Aura.Blog/blob/849eed671fcd5fe6f65e4ca6c70eb11980fb255d/src/Html/Helper/Router.php#L18 and configuration of di here  https://github.com/friendsofaura/Aura.Blog/blob/849eed671fcd5fe6f65e4ca6c70eb11980fb255d/config/Common.php#L55-L56 .

It is nothing new, but all that you guys have discussed above.

**Comment by dlundgren** [comment]
@jakejohns https://gist.github.com/dlundgren/315af5a5aa305c0a349e

While I'm using FoilPHP, the concept is the same to what @harikt has in the Blog helper.

**Comment by dlundgren** [comment]
I forgot to mention that I am doing active link checking by using the urlIs() or urlStartsWithRoute() depending on what navigation needs to be marked as active (parent, child, both).

I think I borrowed the name of the helper from Slim.

**Comment by pmjones** [comment]
I wonder if the thing to do is to include a bare-bones helper in Router, where all it has is a constructor (to inject the Generator) and an __invoke() method (so it is callable). @jakejohns @dlundgren @harikt Thoughts?

**Comment by harikt** [comment]
@pmjones good call, but oh who knows whether the view helper need to implement a special interface? So will that class be helpful to every case scenario? But I agree if it helps, yes we can add.

**Comment by pmjones** [comment]
> who knows whether the view helper need to implement a special interface

(/me nods) Even so, I care mostly about the Aura.\* usage, which only needs it to be a callable.

Other thoughts?

**Comment by harikt** [comment]
:+1:  , ok, then lets add it ;-)  .

**Comment by dlundgren** [comment]
While I feel like Aura.Router shouldn't implement any view helper, I believe it would assist developers to have an implementation that they can at least reference.

**Comment by pmjones** [comment]
@dlundgren Yeah, I get you. I'd like to keep "pollution" down to zero. This seems unobtrusive on its face, and as you say, having the reference point would be good.

Further thoughts?

**Comment by pmjones** [comment]
Anyone feel like putting together a PR for this?

**Comment by jakejohns** [comment]
I made/started using this for now: https://github.com/jnjxp/jnjxp.urlhelper

I figured I'd get some more experience using it before making a pr. Thoughts on what's there so far? It's very experimental.

**Comment by harikt** [comment]
PR https://github.com/auraphp/Aura.Router/pull/129 .

**Comment by pmjones** [comment]
All right, some observations and comments on #129, #130, and #131.

First, the helper under consideration here is not a generic URL helper per se. It is a helper for route URLs only. Anything that's not related to the route need not be in there, so no need for extra query params or trailing strings; the end-user can add those on their own in various ways.

Of the two PRs, then, the more appropriate one is #129. However, #129 needs a little work, which I will detail there.

@harikt Do you get where I'm coming from here?

**Comment by pmjones** [comment]
Fixed by #129.
